### PR TITLE
Ensure correct attributes are always used for embed preview

### PR DESCRIPTION
**Pull request by notnownikki**
## Description

When pasting a YouTube URL multiple times, the responsive styles are not applied on the first render, and so the second render is incorrect and either has scrollbars or whitespace.

This change make sure we are always rendering using the correct attributes, generated from the preview.

## How has this been tested?

Paste a YouTube URL, e.g. https://www.youtube.com/watch?v=wSaoXwQzHnY multiple times into a new post. Each video should be responsive with no large amounts of whitespace or scrollbars.

## Types of changes
Bug fix

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [x] My code follows the accessibility standards. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate.

### CODE REVIEW

**Comment by draganescu** [comment]
Howdy @notnownikki is there a particular browser this happens on, because I tested the way you described and there appears to be no bug on master. 

### On master, after pasting the YouTube PR multiple times I get this:
Screenshot 2019-04-03 at 12 50 42


### On this PR, after pasting the YouTube PR multiple times I get this:
Screenshot 2019-04-03 at 12 51 10

What am I missing and how to test :)?

**Comment by notnownikki** [comment]
I'm using Chrome 73.0.3683.75, the bug _might_ not appear when the window is narrow as you have it there.

Here's what I see on current master, stock WP, 2019 theme, no plugins, browser maximised at 1920x1080 resolution monitor:

whitespace

The wider the window, the worse that whitespace at the bottom gets.

**Comment by notnownikki** [comment]
Also tested in Firefox 66.0.2, and I see the same thing.

**Comment by notnownikki** [comment]
On android chrome, there's a different symptom of the problem, you only see a small section of the video preview. Here's what it looks like pasting a YouTube URL on android chrome, current master:

Screenshot_20190403_111957_com android chrome

This PR fixes that.

**Comment by draganescu** [comment]
Yes @notnownikki I finally could see it happen :D my testing was wrong.

**Comment by draganescu** [comment]
Tested the code and it works as described, there is no more spacing being created for no reason.

File: packages/block-library/src/embed/util.js [hunk 1]
```diff
@@ -178,3 +179,40 @@ export function fallback( url, onReplace ) {
...
+ * Gets block attributes based on the preview and responsive state.
+ *
+ * @param {Object} preview The preview data.
+ * @param {string} title The block's title, e.g. Twitter.
+ * @param {Object} currentClassNames The block's current class names.
+ * @param {boolean} isResponsive Boolean indicating if the block supports responsive content.
+ * @param {boolean} allowResponsive Apply responsive classes to fixed size content.
+ * @return {Object} Attributes and values.
+ */
+export const getAttributesFromPreview = memoize( ( preview, title, currentClassNames, isResponsive, allowResponsive = true ) => {
```
Comment (draganescu) [inline packages/block-library/src/embed/util.js]: Isn't it kinda hard to follow that both the component method and the util function are named the same and then used throughout the code? I wonder if we could rename this one from util, `getMemoizedAttributesFromPreview` ?

File: packages/block-library/src/embed/edit.js [hunk 2]
```diff
@@ -179,6 +152,23 @@ export function getEmbedEditComponent( title, icon, responsive = true ) {
...
+			// the edit component. If we wait for `getAttributesFromPreview` to complete
+			// when the component updates with a preview, we will render an `EmbedPreview`
+			// with incorrect props, and this results in a rendering glitch for responsive
+			// embeds. The glitch happens when the content is made responsive by applying
+			// the responsive styles, but the underlying iframe does not get new HTML
+			// for the sandbox to resize, and so it gets the wrong margins and either
+			// gains a lot of unwanted whitespace, or a scrollbar, depending on the size
+			// and shape of the content. So, we get the attributes from the preview here,
+			// and `getAttributesFromPreview` is memoized in `util` so we're not doing
+			// the computation on every render.
```
Comment (draganescu) [inline packages/block-library/src/embed/edit.js]: Just to verify my understanding, this comment says: on `componentDidUpdate` the preview is not yet ready and hence `handleIncomingPreview` will attempt to `getAttributesFromPreview` from an unfinished preview, so, _as an extra measure_, we call `getAttributesFromPreview` again at render time, and since it is memoized this is very fast. Right?

Comment (notnownikki) [reply packages/block-library/src/embed/edit.js]: Kind of. I think maybe I rambled a little bit in that comment and assumed knowledge I shouldn't have...

The problem is that we need to set attributes based on the preview. If we have that preview when the block is constructed, then there are two renders. The first renders the preview. Then when the attributes are worked out and we call `setAttributes`, the second render applies the responsive classes based on the attributes we got from the preview, and the block's internal state. Applying the classes after the initial render causes the whitespace or scrollbar, because the preview needs to be rendered into a container that already has the correct responsive classes.

So, we always get the attributes from the preview in `render`, because the `setAttributes` call won't have completed on the first render.

We memoize `getAttributesFromPreview` because it requires constructing documents, and we don't want those heavyweight things happening on every render.

I know that's also quite rambling... but hopefully clearer this time? :)

Comment (draganescu) [reply packages/block-library/src/embed/edit.js]: Much better and clearer, it makes sense now that you explained:

> we always get the attributes from the preview in render, because the setAttributes call won't have completed on the first render.

Comment (notnownikki) [reply packages/block-library/src/embed/edit.js]: I'll rewrite that comment so it makes sense to people who are not me :) And I'll also address the `getAttributesFromPreview` naming too. Thank you!

**Comment by notnownikki** [comment]
I have clarified the comment in `render` and renamed the methods that get the attributes from the preview and merge them with the block's current attributes. Hopefully that makes things clearer to new readers :)

**Review by draganescu — APPROVED** [review]
tested again, works great. this is good to go!
